lib/AST/ASTContext.cpp

```diff
@@ -3576,12 +3576,35 @@ void LocatableType::Profile(llvm::FoldingSetNodeID &id, SourceLoc loc,
 // Simple accessors.
 Type ErrorType::get(const ASTContext &C) { return C.TheErrorType; }
 
+static Type replacingTypeVariablesAndPlaceholders(Type ty) {
+  if (!ty || !ty->hasTypeVariableOrPlaceholder())
+    return ty;
+
+  auto &ctx = ty->getASTContext();
+
+  return ty.transformRec([&](Type ty) -> std::optional<Type> {
+    if (!ty->hasTypeVariableOrPlaceholder())
+      return ty;
+
+    // Match the logic in `Solution::simplifyType` and use UnresolvedType.
+    // FIXME: Ideally we'd get rid of UnresolvedType and just use a fresh
+    // PlaceholderType, but we don't currently support placeholders with no
+    // originators.
```

Comment:
Just to note that this is indeed the long-term plan, we need to remove UnresolvedType in favor of PlaceholderType

Reply:
Maybe we should limit this transformation to solver allocated placeholder types instead of all of them. We could also expand ErrorExpr originator and replace type variables with placeholders that point to ASTNode that their locator simplifies down to...

Reply:
> Maybe we should limit this transformation to solver allocated placeholder types instead of all of them

I would rather apply the transform consistently to match what we do in `Solution::simplifyType`, conditionalizing the logic on the allocation arena or originator seems a little odd to me. We should revisit this when we get rid of UnresolvedType though.

> We could also expand ErrorExpr originator and replace type variables with placeholders that point to ASTNode that their locator simplifies down to...

Yeah maybe, I don't think there are any ErrorType consumers that currently really care about originators, but maybe that could be useful for debugging?

Reply:
Yeah, I was mostly talking long-term as part of UnresolvedType removal :)